core/tee/tee_rpmb_fs.c

```diff
@@ -752,6 +752,8 @@ static TEE_Result tee_rpmb_resp_unpack_verify(struct rpmb_data_frame *datafrm,
 	bytes_to_u16(lastfrm.op_result, &op_result);
```

Comment:
@jforissier so with OP-TEE/optee_client#163, `RPMB_RESULT_AUTH_KEY_NOT_PROGRAMMED` (0x07) is eventually passed here as `lastfrm.op_result[1] = 0x00` and `lastfrm.op_result[0] = 0x07`, but this is swapped by `bytes_to_u16` to `0x0700` instead, which returns `TEE_ERROR_GENERIC` below rather than `TEE_ERROR_ITEM_NOT_FOUND`. Can you give some insights please?

Reply:
Changed
`bytes_to_u16(lastfrm.op_result, &op_result);`
to
`op_result = (uint16_t) ((lastfrm.op_result[1] << 8) + lastfrm.op_result[0]);`
OR
`op_result = (uint16_t)lastfrm.op_result[0];`
and was able to pass xtest.
Seems like we don't need to do any swapping of bytes here? Or we needed to swap the bytes in `optee_client`?

Reply:
So I tested with hikey (real rpmb) and the bytes are swapped, which means we need to swap the bytes in `optee_client` before returning?

Reply:
Submitted additional [commit](https://github.com/OP-TEE/optee_client/pull/163/commits/ba613be079121746e6eb8d60234c028f04bccf17) to https://github.com/OP-TEE/optee_client/pull/163 to swap bytes in `optee_client`. Note that this also requires increasing the RPMB size. See [this](https://github.com/OP-TEE/optee_client/pull/163/commits/4d8c2d6e830ff6f190dcc82b9c91bb78f0cdf742) for details.

There's also an alternative [here](https://github.com/vchong/optee_client/commit/da64f089aca0fcd1c60b97ee84d9a314215b3572) which is uglier(?) but more efficient. Also, this one doesn't require the increase in size. Not sure which one you might prefer.

Please review as appropriate. Thanks!

Reply:
`struct rpmb_data_frame` mirrors the eMMC specification (eMMC 4.5 "6.6.22.1 The Data Frame for Replay Protected Memory Block Access"), including byte order, which is specified as "Byte order of the RPMB data frame is MSB first".
This is what the Linux kernel expects to find in the `ioctl()` call, and optee_client / tee-supplicant is not supposed to swap anything (it's just acting as a connection between optee_os and the eMMC device via the Linux driver).

Using the vocabulary from the spec, the  "Operation result" (Table 19) -- 7 bit quantity -- is the LSB of "Operation Results data structure" --16-bit quantity -- minus the high order bit, in other words it is `rpmb_data_frame::op_result[1] & 0x7F` which is probably what we should be doing here instead of `bytes_to_u16()`.

Reply:
In the `optee_client` RPMB emulator, the emulated `ioctl()` calls are just returning one of the below defines. Based on the spec excerpts above, shouldn't it return `(RPMB_RESULT_FOO << 8)` instead if MSB is first? Assume that the write counter status is always ok. This is what I see returned on real devices.
```
 #define RPMB_RESULT_OK                         0x00
 #define RPMB_RESULT_GENERAL_FAILURE            0x01
 #define RPMB_RESULT_AUTH_FAILURE               0x02
 #define RPMB_RESULT_ADDRESS_FAILURE            0x04
```

On the `optee_os` side, I'll add another patch to change `bytes_to_u16()` to `rpmb_data_frame::op_result[1] & 0x7F`.

Reply:
Patch added. Please review PR. Thanks!